zapcore/write_syncer.go

```diff
@@ -75,6 +77,62 @@ func (s *lockedWriteSyncer) Sync() error {
 	return err
 }
 
+type bufferWriterSyncer struct {
+	bufferWriter *bufio.Writer
+	ws           WriteSyncer
+}
+
+// defaultBufferSize sizes the buffer associated with each log file. It's large
+// so that log records can accumulate without the logging thread blocking
+// on disk I/O untill buffer fills up. The flushDaemon will block instead.
+const defaultBufferSize = 256 * 1024
+
+// defaultFlushInterval means the default flush interval
+const defaultFlushInterval = 30 * time.Second
+
+// Buffer wraps a WriteSyncer in a buffer to improve performance,
+func Buffer(ws WriteSyncer, bufferSize int, flushInterval time.Duration) WriteSyncer {
+	if lws, ok := ws.(*lockedWriteSyncer); ok {
+		if _, ok := lws.ws.(*bufferWriterSyncer); ok {
+			// no need to layer on another buffer
+			return ws
+		}
+	}
+
+	if bufferSize == 0 {
+		bufferSize = defaultBufferSize
+	}
+
+	if flushInterval == 0 {
+		flushInterval = defaultFlushInterval
+	}
+
+	// bufio is not goroutine safe, so add lock writer here
+	ws = Lock(&bufferWriterSyncer{
+		bufferWriter: bufio.NewWriterSize(ws, bufferSize),
+	})
+
+	// flush buffer every interval
+	// we do not need exit this goroutine explicitly
+	go func() {
+		for range time.NewTicker(flushInterval).C {
+			if err := ws.Sync(); err != nil {
+				return
+			}
+		}
+	}()
+
+	return ws
+}
+
+func (s *bufferWriterSyncer) Write(bs []byte) (int, error) {
+	return s.bufferWriter.Write(bs)
+}
+
+func (s *bufferWriterSyncer) Sync() error {
+	return s.bufferWriter.Flush()
```

Comment:
I think we also need to call `Sync` on the underlying writer? E.g., when `Fatal` is used, we `Sync()` to make sure the log is written out before returning. If the underlying writer is some network call, then the current implementation's `Flush` will call `Write` on the underlying connection but not flush, so the log may not be emitted.